src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,263 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_integration.h>
+#include <memory>
+
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+/// Numerical integration algorithms
+namespace integration {
+
+namespace detail {
+
+// The GSL functions require the integrand to have this particular function
+// signature. In particular, any extra parameters to the functions must be
+// passed as a void* and re-interpreted appropriately.
+template <typename IntegrandType>
+double integrand(double x, void* params) {
+  const auto function = reinterpret_cast<IntegrandType*>(params);
+  return (*function)(x);
+}
+
+class GslQuadAdaptiveImpl {
+ public:
+  explicit GslQuadAdaptiveImpl(size_t max_intervals) noexcept;
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  // clang-tidy: no runtime references
+  void pup(PUP::er& p) noexcept;  // NOLINT(google-runtime-references)
+
+ private:
+  struct gsl_integration_workspace_deleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept;
+  };
+
+  void initialize() noexcept;
+
+  mutable gsl_function gsl_integrand_;
+
+ protected:
+  template <typename IntegrandType>
+  gsl_function* gsl_integrand(IntegrandType&& integrand) const noexcept {
+    gsl_integrand_.function = &detail::integrand<IntegrandType>;
+    gsl_integrand_.params = &integrand;
+    return &gsl_integrand_;
+  }
+
+  size_t max_intervals_;
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+};
+
+void check_status_code(int status_code);
+
+void disable_gsl_error_handling() noexcept;
+
+}  // namespace detail
+
+/// Defines what algorithm is used and should be chosen according
+/// to the problem.
+enum class GslIntegralType {
+  StandardGaussKronrod,            // gsl_integration_qag()
+  IntegrableSingularitiesPresent,  // gsl_integration_qags()
+  IntegrableSingularitiesKnown,    // gsl_integration_qagp()
+  InfiniteInterval,                // gsl_integration_qagi()
+  UpperBoundaryInfinite,           // gsl_integration_qagiu()
+  LowerBoundaryInfinite            // gsl_integration_qagil()
+};
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A wrapper around the GSL adaptive integration procedures
+ *
+ * All templates take upper bounds to the absolute and relative error;
+ * `tolerance_abs` and `tolerance_rel(default = 0.0)`, respectively.
```

Comment:
I don't know what a relative tolerance of 0.0 means.

Reply:
The documentation mentions "To compute to a specified absolute error, set epsrel to zero. To compute to a specified relative error, set epsabs to zero." If you refer to the meaning of epsabs I would repeat just more of the Introduction section in the documentation here, I'm afraid. ^^

Reply:
Which documentation mentions this? I think it should be explicitly stated here.

Reply:
Ahh, not the documentation in this PR, but the documentation for GSL linked in the description of this PR. Sorry for the confusion.

Reply:
Ah okay, I understand now. So it would be really useful to have `tolerance_rel`/`abs` being zero explained in our documentation because, realistically, it's unlikely people will read our documentation, and basically will never happen that they'll read some other documentation we point to :) (I am sometimes among this group ;) )